### DISCUSSION TITLE
Confirmation Fields Not Validating Consistently

### DISCUSSION THREAD

**Discussion by Coyote6** [livewire/livewire#3118] [7. Bugs (archived)]
### Description
Depending on the order you type into confirmation fields, the validation does not always fire, leaving users with a mismatch error when they are matching or with no error when they are mismatched.

### Exact steps to reproduce
1. Type 'Bob' into the confirmation field first.
2. Then type 'Bob' into the first field.  At this point validation works. 
3. Continue typing 'by' (making the name Bobby) 
4. Then go back to the confirmation field, and type 'by'. At this point both fields should read 'Bobby', but the confirm field no longer validates, leaving you with a mismatch error.
5. Delete the 'by' from the top field.
6. Delete the 'by' from the confirm field. At this point both fields should read 'Bob', but the confirm field still leaves the error.
7. Add 'by' back to the top field.
8. Add 'by' back to the confirm field. Error still shows.
9. Remove 'by' from the confirm field first.
10. Remove 'by' from the top field.  Error no longer shows.

### Stripped-down, copy-pastable code snippets
In the component
```
namespace App\Http\Livewire\Auth;
use Livewire\Component;

class Test extends Component {
	
  public $name = '';
  public $nameConfirmation = '';

  protected function rules() {
      return [
      	"name" => [
            "string",
            "required",
            "same:nameConfirmation"
        ],	
        "nameConfirmation" => [
            "string"
         ]
      ];
  }
	
  public function updated ($field) {
      $this->validateOnly($field);
  }
			
  public function register () {
      $values = $this->validate();
      return $values;
  }
		
  public function render() {
      return view('livewire.auth.test');
  }
    
}
```

In the blade template:
```
<form wire:submit.prevent="register" method="post" action="/register">
	<input class="border mb-2 border-gray-300 block" wire:model="name" required name="name" type="text" placeholder="Name">
	@error ('name')
		{{ $message }}
	@enderror
	<input class="border mb-2 border-gray-300 block" wire:model="nameConfirmation" name="nameConfirmation" type="text" placeholder="Confirm Name" >
	@error ('nameConfirmation')
		{{ $message }}
	@enderror
	<button class="border mb-2 border-gray-300 block" id="submit">Register</button>
	@csrf
</form>
```
**Context**
- Livewire version: v2.3.18
- Laravel version: 8.28.1
- Alpine version: unknown
- Browser: Chrome, Safari, Firefox, Opera - Didn't check edge

**Comment by Coyote6** [comment]
Still need to test some more, but a possible fix is to add a script to update the main input on keyup and blur events from the confirmation field.  I feel like this would be even nicer if it was out of the box, but this is what I currently have.

In you base app view where @livewireScripts is called, add:
```
@stack('scripts')
```

In the view for the component:
1. Add an id attribute to the main input:
```
<input class="border mb-2 border-gray-300 block" wire:model="name" required id="name" name="name" type="text" placeholder="Name">
```
2. Add wire:keyup and wire:blur events to confirmation input:
```
<input class="border mb-2 border-gray-300 block" wire:model="nameConfirmation" name="nameConfirmation" type="text" placeholder="Confirm Name" wire:keyup="$emit('updatedConfirmation', 'name')" wire:blur="$emit('updatedConfirmation', 'name')">
```

3. Add a script that is pushed to the scripts stack.
```
@push('scripts')
	<script>
		Livewire.on('updatedConfirmation', name => {
			var el = document.getElementById(name);
			var	val = el.value;
			@this.set(name, val);
		});
	</script>
@endpush
````

The final component view should look something like this:
```
<form wire:submit.prevent="register" method="post" action="/register">
	<input class="border mb-2 border-gray-300 block" wire:model="name" required id="name" name="name" type="text" placeholder="Name">
	@error ('name')
		{{ $message }}
	@enderror
	<input class="border mb-2 border-gray-300 block" wire:model="nameConfirmation" name="nameConfirmation" type="text" placeholder="Confirm Name" wire:keyup="$emit('updatedConfirmation', 'name')" wire:blur="$emit('updatedConfirmation', 'name')">
	@error ('nameConfirmation')
		{{ $message }}
	@enderror
	<button class="border mb-2 border-gray-300 block" id="submit">Register</button>
	@csrf
	@push('scripts')
		<script>
			Livewire.on('updatedConfirmation', name => {
				var el = document.getElementById(name);
				var	val = el.value;
				@this.set(name, val);
			});
		</script>
	@endpush
</form>
```

**Comment by joshhanley** [comment]
@Coyote6 looking at your original issue, I believe the issue is you are only validating one field at a time, by using validateOnly.

If they should be validated together, regardless of which one has been input then you should change your `updated` method, to `$this->validate();`.

Also, not sure if you are aware, but there is a `confirmed` validation rule, check it out here https://laravel.com/docs/8.x/validation#rule-confirmed

Hope this helps!

**Comment by Coyote6** [comment]
If you use `$this->validate()` yes it solves that particular issue, but say you had another field below that is required, that the user hasn't got to yet.  It causes it to show an error even though the user has not reached that point in the form.
```
namespace App\Http\Livewire\Auth;

use Livewire\Component;

class Test extends Component
{
	
	public $name = '';
	public $nameConfirmation = '';
	public $newField = '';

		
	protected function rules() {
		 return [
			"name" => [
				"string",
				"required",
				"confirmed"
			],	
			"nameConfirmation" => [
				"string"
			],
			'newField' => [
				'required'
			]
 		];
	}
    
		
	public function updated ($field) {
		$this->validate();
	}
	
		
	public function register () {
		$values = $this->validate();
		return $values;
	}
		
		
	public function render() {
		return view('livewire.auth.test');
 	}
    
    
 }
```
```
<form wire:submit.prevent="register" method="post" action="/register">
	<input class="border mb-2 border-gray-300 block" wire:model="name" required id="name" name="name" type="text" placeholder="Name">
	@error ('name')
		{{ $message }}
	@enderror
	<input class="border mb-2 border-gray-300 block" wire:model="nameConfirmation" name="nameConfirmation" type="text" placeholder="Confirm Name">
	@error ('nameConfirmation')
		{{ $message }}
	@enderror
	<input class="border mb-2 border-gray-300 block" wire:model="newField" name="newField" type="text" placeholder="New Field">
	@error ('newField')
		{{ $message }}
	@enderror
	<button class="border mb-2 border-gray-300 block" id="submit">Register</button>
	@csrf
</form>
```

Yes, I use `confirmed` validation most the time.  It does however make you name the confirmation field with `_confirmation` as the name, meaning your property name would need to be snake cased, and I was just following along with the Livewire tutorial I found in which he used `same:`.  I am still very new to Livewire, so there very well maybe a better way.  I just think confirm fields need to be handled at the same time when validateOnly() is called, otherwise the errors will never be correct.
